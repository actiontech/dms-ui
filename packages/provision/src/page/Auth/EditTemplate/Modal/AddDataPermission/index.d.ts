export interface IAddDataPermission {
  dataPermissions: IDataPermissionsTable[];
  setDataPermissions: (permissions: IDataPermissionsTable[]) => void;
  editIndex?: number;
  setEditIndex: (index?: number) => void;
}

export interface IFormFields {
  business?: string;
  service?: string;
  data_objects?: IDataObjects[];
  data_operations?: string[];
}

export interface IDataPermissionsTable {
  index: number;
  business: string;
  serviceValue: string;
  objectsValue?: IDataObjects[];
  operationsValue: string[];

  serviceLabel: string;
  objectsLabel: string[];
  operationsLabel: string[];

  objectsParams: string[];
}

export interface IDataObjects {
  database?: string;
  tables?: string[];
}
