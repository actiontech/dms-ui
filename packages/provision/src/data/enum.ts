export enum ModalName {
  GetConnection = 'GET_CONNECTION',

  UpdateTemplateInAuth = 'UPDATE_TEMPLATE_IN_AUTH',

  UpdateUserInAuth = 'UPDATE_USER_IN_AUTH',

  UpdateExpirationInAuth = 'UPDATE_EXPIRATION_IN_AUTH',

  DataPermissionModal = 'DATA_PERMISSION_MODAL',

  CopyTemplate = 'COPY_TEMPLATE'
}

export enum EventEmitterKey {
  Refresh_Auth_List_Table = 'Refresh_Auth_List_Table',

  Refresh_Auth_Template_List_Table = 'Refresh_Auth_Template_List_Table'
}
