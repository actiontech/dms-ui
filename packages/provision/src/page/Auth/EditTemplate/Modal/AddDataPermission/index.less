.add-data-permission-form {
  .anticon-plus-circle,
  .anticon-minus-circle {
    margin-left: 5px;
    line-height: 32px;
  }
}
