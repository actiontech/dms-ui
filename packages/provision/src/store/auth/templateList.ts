import { atom } from 'recoil';
import { ModalStatus } from '~/types/common.type';
import { StoreKey } from '..';
import { IListDataPermissionTemplate } from '@actiontech/shared/lib/api/provision/service/common';

export const AuthDataPermissionListModalStatus = atom<ModalStatus>({
  key: StoreKey.Auth_Data_Permission_List_Modal_Status,
  default: {}
});

export const AuthTemplateModalStatus = atom<ModalStatus>({
  key: StoreKey.Auth_Template_Modal_Status,
  default: {}
});

export const AuthTemplateListSelectData =
  atom<IListDataPermissionTemplate | null>({
    key: StoreKey.Auth_Template_List_Selected_Data,
    default: null
  });
