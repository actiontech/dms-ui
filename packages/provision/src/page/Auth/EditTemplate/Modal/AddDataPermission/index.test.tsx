import {
  act,
  cleanup,
  fireEvent,
  screen,
  waitFor
} from '@testing-library/react';
import { ModalName } from '~/data/enum';
import { AuthDataPermissionListModalStatus } from '~/store/auth/templateList';
import { getAllBySelector, getBySelector } from '~/testUtil/customQuery';
import AddDataPermission from '.';
import { IAddDataPermission } from './index.type';
import auth from '~/testUtil/mockApi/auth';
import { selectOptionByIndex } from '@actiontech/shared/lib/testUtil/customQuery';
import { mockUseCurrentProject } from '@actiontech/shared/lib/testUtil/mockHook/mockUseCurrentProject';
import { superRender } from '@actiontech/shared/lib/testUtil/customRender';

const defaultDataPermissions = [
  {
    serviceLabel: 'Julian Lueilwitz',
    objectsLabel: ['database-2.*'],
    operationsLabel: ['查询'],

    business: 'business-1',
    serviceValue: '42343',
    objectsValue: [{ database: '2', tables: undefined }],
    operationsValue: ['72999044'],

    objectsParams: []
  }
];

describe('/Auth/AddDataPermission', () => {
  let listBusinessesSpy: jest.SpyInstance;
  let listServicesSpy: jest.SpyInstance;
  let listDataBasesSpy: jest.SpyInstance;
  let listTablesSpy: jest.SpyInstance;
  let listOperationSetsSpy: jest.SpyInstance;

  beforeEach(() => {
    mockUseCurrentProject();
    listBusinessesSpy = auth.listBusinesses();
    listServicesSpy = auth.listServices();
    listDataBasesSpy = auth.listDataBases();
    listTablesSpy = auth.listTables();
    listOperationSetsSpy = auth.listOperationSets();

    jest.useFakeTimers();
    window.confirm = jest.fn();
  });
  afterEach(() => {
    jest.useRealTimers();
    jest.clearAllTimers();
    jest.clearAllMocks();
    cleanup();
  });
  const customRender = (
    defaultVisible = true,
    props?: Partial<IAddDataPermission>
  ) => {
    return superRender(
      <AddDataPermission
        dataPermissions={defaultDataPermissions}
        setDataPermissions={jest.fn()}
        setEditIndex={jest.fn()}
        {...props}
      />,
      {},
      {
        recoilRootProps: {
          initializeState({ set }) {
            set(AuthDataPermissionListModalStatus, {
              [ModalName.DataPermissionModal]: defaultVisible
            });
          }
        }
      }
    );
  };
  it('should match close modal when modal status is false', async () => {
    const { container } = customRender(false);
    expect(container).toMatchInlineSnapshot(`<div />`);
  });

  it('should match open modal when modal status is true', async () => {
    const { baseElement } = customRender();
    await screen.findByText('添加数据权限');

    expect(baseElement).toMatchSnapshot();
  });

  it('this option should be disabled and update table list when select database', async () => {
    customRender(true, {
      dataPermissions: [
        {
          serviceLabel: 'Julian Lueilwitz',
          objectsLabel: ['*'],
          operationsLabel: ['查询'],

          business: 'business-1',
          serviceValue: '42343',
          objectsValue: [{}],
          operationsValue: ['72999044'],

          objectsParams: []
        }
      ],
      editIndex: 0
    });

    await act(async () => jest.advanceTimersByTime(3000));
    selectOptionByIndex('业务', 'business-2', 1);
    await act(async () => jest.advanceTimersByTime(300));
    expect(listServicesSpy).toHaveBeenCalled();
    await act(async () => jest.advanceTimersByTime(3000));
    fireEvent.mouseDown(
      getBySelector('.data-service-select .ant-select-selector input')
    );
    await act(async () => jest.advanceTimersByTime(300));
    await act(() =>
      fireEvent.click(screen.getAllByText('Julian Lueilwitz(MySQL)')[0])
    );
    await act(async () => jest.advanceTimersByTime(300));
    expect(listDataBasesSpy).toHaveBeenCalled();
    await act(async () => jest.advanceTimersByTime(3000));
    fireEvent.mouseDown(
      getBySelector(
        '.ant-select-selection-search-input',
        screen.getByTestId('database-form-item-database-0')
      )
    );
    expect(screen.getByTestId('database-form-item-database-0')).toHaveClass(
      'ant-select-open'
    );
    await screen.findByText('database-1');
    await act(() => fireEvent.click(screen.getByText('database-1')));
    expect(listTablesSpy).toHaveBeenCalled();
    await act(async () => jest.advanceTimersByTime(3000));
    fireEvent.mouseDown(
      getBySelector(
        '.ant-select-selection-search-input',
        screen.getByTestId('database-form-item-schema-0')
      )
    );
    await act(async () => jest.advanceTimersByTime(300));
    expect(screen.queryByText('table-1')).toBeInTheDocument();
  });

  it('should push dataPermissions when click add button', async () => {
    const setDataPermissions = jest.fn();
    customRender(true, {
      setDataPermissions,
      editIndex: 0
    });
    await act(async () => jest.advanceTimersByTime(3000));
    fireEvent.click(screen.getByText('提 交'));
    expect(setDataPermissions).toHaveBeenCalledTimes(0);
    expect(listDataBasesSpy).toHaveBeenCalledTimes(1);
    await screen.findByText(
      '已存在相同数据源、相同数据对象、相同数据操作的权限'
    );
    await act(async () => jest.advanceTimersByTime(200));

    fireEvent.mouseDown(
      getBySelector(
        '.ant-select-selection-search-input',
        screen.getByTestId('database-form-item-database-0')
      )
    );
    await act(async () => jest.advanceTimersByTime(100));
    fireEvent.click(screen.getAllByText('database-2')[1]);
    await act(async () => jest.advanceTimersByTime(300));
    await act(async () => jest.advanceTimersByTime(3000));
    fireEvent.mouseDown(
      getBySelector(
        '.ant-select-selection-search-input',
        screen.getByTestId('database-form-item-schema-0')
      )
    );
    await act(async () => jest.advanceTimersByTime(300));
    await act(() => fireEvent.click(screen.getByText('table-1')));
    await act(() => fireEvent.click(getBySelector('.add-object-button')));
    fireEvent.mouseDown(
      getBySelector(
        '.ant-select-selection-search-input',
        screen.getByTestId('database-form-item-database-1')
      )
    );
    await act(() => fireEvent.click(screen.getAllByText('database-2')[1]));
    await act(async () => jest.advanceTimersByTime(300));
    await act(async () => jest.advanceTimersByTime(3000));
    selectOptionByIndex('选择操作', '查询', 0);

    await act(() => fireEvent.click(screen.getByText('提 交')));
    expect(setDataPermissions).toHaveBeenCalledTimes(1);
  });

  it('remove table options when click remove database field icon', async () => {
    customRender(true, {
      editIndex: 0
    });
    await act(async () => jest.advanceTimersByTime(3000));
    fireEvent.mouseDown(
      getBySelector(
        '.ant-select-selection-search-input',
        screen.getByTestId('database-form-item-schema-0')
      )
    );
    await act(async () => jest.advanceTimersByTime(300));
    await act(async () => jest.advanceTimersByTime(3000));
    const el = await screen.findByText('table-1');
    expect(el).toBeInTheDocument();
    fireEvent.click(getBySelector('.ant-select-clear'));
    await act(async () => jest.advanceTimersByTime(300));
  });

  it('remove data_objects field when click remove field button', async () => {
    customRender();
    fireEvent.click(getBySelector('.add-object-button'));
    const objectFieldsDom = getAllBySelector('.database-form-item');
    expect(objectFieldsDom.length).toBe(2);
    fireEvent.click(getBySelector('.remove-object-button'));
    const newObjectFieldsDom = getAllBySelector('.database-form-item');
    expect(newObjectFieldsDom.length).toBe(1);
  });
});
