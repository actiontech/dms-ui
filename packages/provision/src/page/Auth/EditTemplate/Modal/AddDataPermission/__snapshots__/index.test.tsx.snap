// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/Auth/AddDataPermission should match open modal when modal status is true 1`] = `
<body>
  <div />
  <div>
    <div
      class="ant-drawer ant-drawer-right css-dev-only-do-not-override-txh9fw ant-drawer-open"
      tabindex="-1"
    >
      <div
        class="ant-drawer-mask"
      />
      <div
        aria-hidden="true"
        data-sentinel="start"
        style="width: 0px; height: 0px; overflow: hidden; outline: none; position: absolute;"
        tabindex="0"
      />
      <div
        class="ant-drawer-content-wrapper"
        style="width: 640px;"
      >
        <div
          aria-modal="true"
          class="ant-drawer-content basic-drawer-wrapper css-1q29jje"
          role="dialog"
        >
          <div
            class="ant-drawer-wrapper-body"
          >
            <div
              class="ant-drawer-header"
            >
              <div
                class="ant-drawer-header-title"
              >
                <div
                  class="ant-drawer-title"
                >
                  添加数据权限
                </div>
              </div>
            </div>
            <div
              class="ant-drawer-body"
            >
              <div
                class="closed-icon-custom"
              >
                <span
                  aria-label="close"
                  class="anticon anticon-close"
                  role="img"
                >
                  <svg
                    aria-hidden="true"
                    data-icon="close"
                    fill="currentColor"
                    focusable="false"
                    height="1em"
                    viewBox="64 64 896 896"
                    width="1em"
                  >
                    <path
                      d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                    />
                  </svg>
                </span>
              </div>
              <form
                class="ant-form ant-form-vertical css-dev-only-do-not-override-txh9fw add-data-permission-form"
              >
                <div
                  class="ant-form-item css-dev-only-do-not-override-txh9fw"
                >
                  <div
                    class="ant-row ant-form-item-row css-dev-only-do-not-override-txh9fw"
                  >
                    <div
                      class="ant-col ant-form-item-label css-dev-only-do-not-override-txh9fw"
                    >
                      <label
                        class="ant-form-item-required"
                        for="business"
                        title="业务"
                      >
                        业务
                      </label>
                    </div>
                    <div
                      class="ant-col ant-form-item-control css-dev-only-do-not-override-txh9fw"
                    >
                      <div
                        class="ant-form-item-control-input"
                      >
                        <div
                          class="ant-form-item-control-input-content"
                        >
                          <div
                            aria-required="true"
                            class="ant-select ant-select-in-form-item basic-select-wrapper css-iy4o4x css-dev-only-do-not-override-v9sov3 ant-select-single ant-select-show-arrow ant-select-loading"
                          >
                            <div
                              class="ant-select-selector"
                            >
                              <span
                                class="ant-select-selection-search"
                              >
                                <input
                                  aria-activedescendant="business_list_0"
                                  aria-autocomplete="list"
                                  aria-controls="business_list"
                                  aria-expanded="false"
                                  aria-haspopup="listbox"
                                  aria-owns="business_list"
                                  aria-required="true"
                                  autocomplete="off"
                                  class="ant-select-selection-search-input"
                                  id="business"
                                  readonly=""
                                  role="combobox"
                                  style="opacity: 0;"
                                  type="search"
                                  unselectable="on"
                                  value=""
                                />
                              </span>
                              <span
                                class="ant-select-selection-placeholder"
                              >
                                请选择{{name}}
                              </span>
                            </div>
                            <span
                              aria-hidden="true"
                              class="ant-select-arrow ant-select-arrow-loading"
                              style="user-select: none;"
                              unselectable="on"
                            >
                              <span
                                aria-label="loading"
                                class="anticon anticon-loading anticon-spin"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="loading"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="0 0 1024 1024"
                                  width="1em"
                                >
                                  <path
                                    d="M988 548c-19.9 0-36-16.1-36-36 0-59.4-11.6-117-34.6-171.3a440.45 440.45 0 00-94.3-139.9 437.71 437.71 0 00-139.9-94.3C629 83.6 571.4 72 512 72c-19.9 0-36-16.1-36-36s16.1-36 36-36c69.1 0 136.2 13.5 199.3 40.3C772.3 66 827 103 874 150c47 47 83.9 101.8 109.7 162.7 26.7 63.1 40.2 130.2 40.2 199.3.1 19.9-16 36-35.9 36z"
                                  />
                                </svg>
                              </span>
                            </span>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="ant-form-item css-dev-only-do-not-override-txh9fw"
                >
                  <div
                    class="ant-row ant-form-item-row css-dev-only-do-not-override-txh9fw"
                  >
                    <div
                      class="ant-col ant-form-item-label css-dev-only-do-not-override-txh9fw"
                    >
                      <label
                        class="ant-form-item-required"
                        for="service"
                        title="选择数据源"
                      >
                        选择数据源
                      </label>
                    </div>
                    <div
                      class="ant-col ant-form-item-control css-dev-only-do-not-override-txh9fw"
                    >
                      <div
                        class="ant-form-item-control-input"
                      >
                        <div
                          class="ant-form-item-control-input-content"
                        >
                          <div
                            aria-required="true"
                            class="ant-row ant-row-no-wrap css-dev-only-do-not-override-txh9fw"
                            id="service"
                          >
                            <div
                              class="ant-col css-dev-only-do-not-override-txh9fw"
                              style="flex: 1 1 auto; min-width: 0;"
                            >
                              <div
                                aria-required="true"
                                class="ant-select ant-select-in-form-item basic-select-wrapper data-service-select css-iy4o4x css-dev-only-do-not-override-v9sov3 ant-select-single ant-select-show-arrow"
                              >
                                <div
                                  class="ant-select-selector"
                                >
                                  <span
                                    class="ant-select-selection-search"
                                  >
                                    <input
                                      aria-activedescendant="service_list_0"
                                      aria-autocomplete="list"
                                      aria-controls="service_list"
                                      aria-expanded="false"
                                      aria-haspopup="listbox"
                                      aria-owns="service_list"
                                      aria-required="true"
                                      autocomplete="off"
                                      class="ant-select-selection-search-input"
                                      id="service"
                                      readonly=""
                                      role="combobox"
                                      style="opacity: 0;"
                                      type="search"
                                      unselectable="on"
                                      value=""
                                    />
                                  </span>
                                  <span
                                    class="ant-select-selection-placeholder"
                                  >
                                    请选择{{name}}
                                  </span>
                                </div>
                                <span
                                  aria-hidden="true"
                                  class="ant-select-arrow"
                                  style="user-select: none;"
                                  unselectable="on"
                                >
                                  <span
                                    aria-label="down"
                                    class="anticon anticon-down ant-select-suffix"
                                    role="img"
                                  >
                                    <svg
                                      aria-hidden="true"
                                      data-icon="down"
                                      fill="currentColor"
                                      focusable="false"
                                      height="1em"
                                      viewBox="64 64 896 896"
                                      width="1em"
                                    >
                                      <path
                                        d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                                      />
                                    </svg>
                                  </span>
                                </span>
                              </div>
                            </div>
                            <div
                              class="ant-col css-dev-only-do-not-override-txh9fw"
                              style="flex: 0 0 48px; min-width: 0;"
                            >
                              <div
                                class="ant-space css-dev-only-do-not-override-txh9fw ant-space-horizontal ant-space-align-center basic-tooltips-wrapper css-4jbtl9"
                              >
                                <div
                                  class="ant-space-item"
                                >
                                  <button
                                    class="ant-btn css-dev-only-do-not-override-674gwq ant-btn-default ant-btn-icon-only basic-button-wrapper css-iieact"
                                    disabled=""
                                    style="margin-left: 12px;"
                                    type="button"
                                  >
                                    <span
                                      class="ant-btn-icon"
                                    >
                                      <span
                                        class="custom-icon css-u3jcp4"
                                      >
                                        <svg
                                          fill="none"
                                          height="16"
                                          viewBox="0 0 16 16"
                                          width="16"
                                          xmlns="http://www.w3.org/2000/svg"
                                        >
                                          <path
                                            d="M14.6666 6.93337V2.20004C14.6666 1.80004 14.3333 1.46671 13.9333 1.46671H8.99992C8.59992 1.46671 8.26658 1.80004 8.26658 2.20004V6.93337C8.26658 7.33337 8.59992 7.66671 8.99992 7.66671H13.9333C14.3333 7.66671 14.6666 7.33337 14.6666 6.93337ZM2.06659 8.40004C1.66659 8.40004 1.33325 8.73337 1.33325 9.13337V13.8667C1.33325 14.2667 1.66659 14.6 2.06659 14.6H6.99992C7.39992 14.6 7.73325 14.2667 7.73325 13.8667V9.13337C7.73325 8.73337 7.39992 8.40004 6.99992 8.40004H2.06659ZM4.79992 5.20004H4.06659C4.06659 3.93337 5.13325 2.86671 6.39992 2.86671H6.66658C6.86658 2.86671 6.99992 2.73337 6.99992 2.53337V1.66671C6.99992 1.46671 6.86658 1.33337 6.66658 1.33337H6.39992C4.26659 1.33337 2.53325 3.06671 2.53325 5.20004H1.66659C1.53325 5.20004 1.46659 5.33337 1.46659 5.46671C1.46659 5.53337 1.46659 5.53337 1.53325 5.60004L2.99992 7.26671C3.13325 7.40004 3.33325 7.40004 3.46659 7.26671L4.93325 5.60004C4.99992 5.53337 4.99992 5.40004 4.93325 5.33337C4.86659 5.26671 4.86659 5.20004 4.79992 5.20004ZM11.1999 10.8H11.9333C11.9333 12.0667 10.8666 13.1334 9.59992 13.1334H9.33325C9.13325 13.1334 8.99992 13.2667 8.99992 13.4667V14.4C8.99992 14.6 9.13325 14.7334 9.33325 14.7334H9.59992C11.7333 14.7334 13.4666 13 13.4666 10.8667H14.3333C14.4666 10.8667 14.5333 10.8 14.5333 10.6667C14.5333 10.6 14.5333 10.6 14.4666 10.5334L12.9999 8.86671C12.8666 8.73337 12.6666 8.73337 12.5333 8.86671L11.0666 10.5334C10.9999 10.6 10.9999 10.7334 11.0666 10.8C11.0666 10.7334 11.1333 10.8 11.1999 10.8Z"
                                            fill="currentColor"
                                          />
                                        </svg>
                                      </span>
                                    </span>
                                  </button>
                                </div>
                              </div>
                            </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="ant-form-item database-form-item css-dev-only-do-not-override-txh9fw"
                >
                  <div
                    class="ant-row ant-form-item-row css-dev-only-do-not-override-txh9fw"
                  >
                    <div
                      class="ant-col ant-form-item-label css-dev-only-do-not-override-txh9fw"
                    >
                      <label
                        class=""
                        title="选择数据库表"
                      >
                        选择数据库表
                      </label>
                    </div>
                    <div
                      class="ant-col ant-form-item-control css-dev-only-do-not-override-txh9fw"
                    >
                      <div
                        class="ant-form-item-control-input"
                      >
                        <div
                          class="ant-form-item-control-input-content"
                        >
                          <div
                            class="ant-row ant-row-no-wrap css-dev-only-do-not-override-txh9fw"
                          >
                            <div
                              class="ant-col css-dev-only-do-not-override-txh9fw"
                              style="flex: 0 0 180px; min-width: 0; margin-right: 12px;"
                            >
                              <div
                                class="ant-select ant-select-in-form-item basic-select-wrapper css-iy4o4x css-dev-only-do-not-override-v9sov3 ant-select-single ant-select-allow-clear ant-select-show-arrow"
                                data-testid="database-form-item-database-0"
                              >
                                <div
                                  class="ant-select-selector"
                                >
                                  <span
                                    class="ant-select-selection-search"
                                  >
                                    <input
                                      aria-activedescendant="data_objects_0_database_list_0"
                                      aria-autocomplete="list"
                                      aria-controls="data_objects_0_database_list"
                                      aria-expanded="false"
                                      aria-haspopup="listbox"
                                      aria-owns="data_objects_0_database_list"
                                      autocomplete="off"
                                      class="ant-select-selection-search-input"
                                      id="data_objects_0_database"
                                      readonly=""
                                      role="combobox"
                                      style="opacity: 0;"
                                      type="search"
                                      unselectable="on"
                                      value=""
                                    />
                                  </span>
                                  <span
                                    class="ant-select-selection-placeholder"
                                  >
                                    *
                                  </span>
                                </div>
                                <span
                                  aria-hidden="true"
                                  class="ant-select-arrow"
                                  style="user-select: none;"
                                  unselectable="on"
                                >
                                  <span
                                    aria-label="down"
                                    class="anticon anticon-down ant-select-suffix"
                                    role="img"
                                  >
                                    <svg
                                      aria-hidden="true"
                                      data-icon="down"
                                      fill="currentColor"
                                      focusable="false"
                                      height="1em"
                                      viewBox="64 64 896 896"
                                      width="1em"
                                    >
                                      <path
                                        d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                                      />
                                    </svg>
                                  </span>
                                </span>
                              </div>
                            </div>
                            <div
                              class="ant-col css-dev-only-do-not-override-txh9fw"
                              style="flex: 1 1 auto; min-width: 0;"
                            >
                              <div
                                class="ant-select ant-select-in-form-item basic-select-wrapper css-iy4o4x css-dev-only-do-not-override-v9sov3 ant-select-multiple ant-select-show-arrow ant-select-show-search"
                                data-testid="database-form-item-schema-0"
                              >
                                <div
                                  class="ant-select-selector"
                                >
                                  <div
                                    class="ant-select-selection-overflow"
                                  >
                                    <div
                                      class="ant-select-selection-overflow-item ant-select-selection-overflow-item-suffix"
                                      style="opacity: 1;"
                                    >
                                      <div
                                        class="ant-select-selection-search"
                                        style="width: 0px;"
                                      >
                                        <input
                                          aria-activedescendant="data_objects_0_tables_list_0"
                                          aria-autocomplete="list"
                                          aria-controls="data_objects_0_tables_list"
                                          aria-expanded="false"
                                          aria-haspopup="listbox"
                                          aria-owns="data_objects_0_tables_list"
                                          autocomplete="off"
                                          class="ant-select-selection-search-input"
                                          id="data_objects_0_tables"
                                          readonly=""
                                          role="combobox"
                                          style="opacity: 0;"
                                          type="search"
                                          unselectable="on"
                                          value=""
                                        />
                                        <span
                                          aria-hidden="true"
                                          class="ant-select-selection-search-mirror"
                                        >
                                           
                                        </span>
                                      </div>
                                    </div>
                                  </div>
                                  <span
                                    class="ant-select-selection-placeholder"
                                  >
                                    *
                                  </span>
                                </div>
                                <span
                                  aria-hidden="true"
                                  class="ant-select-arrow"
                                  style="user-select: none;"
                                  unselectable="on"
                                >
                                  <span
                                    aria-label="down"
                                    class="anticon anticon-down ant-select-suffix"
                                    role="img"
                                  >
                                    <svg
                                      aria-hidden="true"
                                      data-icon="down"
                                      fill="currentColor"
                                      focusable="false"
                                      height="1em"
                                      viewBox="64 64 896 896"
                                      width="1em"
                                    >
                                      <path
                                        d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                                      />
                                    </svg>
                                  </span>
                                </span>
                              </div>
                            </div>
                            <div
                              class="ant-col css-dev-only-do-not-override-txh9fw"
                              style="flex: 0 0 48px; min-width: 0;"
                            />
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="ant-form-item css-dev-only-do-not-override-txh9fw"
                >
                  <div
                    class="ant-row ant-form-item-row css-dev-only-do-not-override-txh9fw"
                  >
                    <div
                      class="ant-col ant-form-item-control css-dev-only-do-not-override-txh9fw"
                    >
                      <div
                        class="ant-form-item-control-input"
                      >
                        <div
                          class="ant-form-item-control-input-content"
                        >
                          <button
                            class="ant-btn css-dev-only-do-not-override-674gwq ant-btn-default form-list-add add-object-button basic-button-wrapper css-1719g3l"
                            type="button"
                          >
                            <span
                              class="ant-btn-icon"
                            >
                              <svg
                                fill="none"
                                height="16"
                                viewBox="0 0 16 16"
                                width="16"
                                xmlns="http://www.w3.org/2000/svg"
                              >
                                <path
                                  d="M7.00016 13.6667C3.31816 13.6667 0.333496 10.682 0.333496 7.00004C0.333496 3.31804 3.31816 0.333374 7.00016 0.333374C10.6822 0.333374 13.6668 3.31804 13.6668 7.00004C13.6668 10.682 10.6822 13.6667 7.00016 13.6667ZM6.3335 6.33337H3.66683V7.66671H6.3335V10.3334H7.66683V7.66671H10.3335V6.33337H7.66683V3.66671H6.3335V6.33337Z"
                                  fill="#4583FF"
                                />
                              </svg>
                            </span>
                            <span>
                              添加数据库表
                            </span>
                          </button>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="ant-form-item css-dev-only-do-not-override-txh9fw"
                >
                  <div
                    class="ant-row ant-form-item-row css-dev-only-do-not-override-txh9fw"
                  >
                    <div
                      class="ant-col ant-form-item-label css-dev-only-do-not-override-txh9fw"
                    >
                      <label
                        class="ant-form-item-required"
                        for="data_operations"
                        title="选择操作"
                      >
                        选择操作
                      </label>
                    </div>
                    <div
                      class="ant-col ant-form-item-control css-dev-only-do-not-override-txh9fw"
                    >
                      <div
                        class="ant-form-item-control-input"
                      >
                        <div
                          class="ant-form-item-control-input-content"
                        >
                          <div
                            aria-required="true"
                            class="ant-select ant-select-in-form-item basic-select-wrapper css-iy4o4x css-dev-only-do-not-override-v9sov3 ant-select-multiple ant-select-show-arrow ant-select-show-search"
                          >
                            <div
                              class="ant-select-selector"
                            >
                              <div
                                class="ant-select-selection-overflow"
                              >
                                <div
                                  class="ant-select-selection-overflow-item ant-select-selection-overflow-item-suffix"
                                  style="opacity: 1;"
                                >
                                  <div
                                    class="ant-select-selection-search"
                                    style="width: 0px;"
                                  >
                                    <input
                                      aria-activedescendant="data_operations_list_0"
                                      aria-autocomplete="list"
                                      aria-controls="data_operations_list"
                                      aria-expanded="false"
                                      aria-haspopup="listbox"
                                      aria-owns="data_operations_list"
                                      aria-required="true"
                                      autocomplete="off"
                                      class="ant-select-selection-search-input"
                                      id="data_operations"
                                      readonly=""
                                      role="combobox"
                                      style="opacity: 0;"
                                      type="search"
                                      unselectable="on"
                                      value=""
                                    />
                                    <span
                                      aria-hidden="true"
                                      class="ant-select-selection-search-mirror"
                                    >
                                       
                                    </span>
                                  </div>
                                </div>
                              </div>
                              <span
                                class="ant-select-selection-placeholder"
                              >
                                请选择{{name}}
                              </span>
                            </div>
                            <span
                              aria-hidden="true"
                              class="ant-select-arrow"
                              style="user-select: none;"
                              unselectable="on"
                            >
                              <span
                                aria-label="down"
                                class="anticon anticon-down ant-select-suffix"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="down"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="64 64 896 896"
                                  width="1em"
                                >
                                  <path
                                    d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                                  />
                                </svg>
                              </span>
                            </span>
                          </div>
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </form>
            </div>
            <div
              class="ant-drawer-footer"
            >
              <div
                class="ant-space css-dev-only-do-not-override-txh9fw ant-space-horizontal ant-space-align-center"
              >
                <div
                  class="ant-space-item"
                  style="margin-right: 8px;"
                >
                  <button
                    class="ant-btn css-dev-only-do-not-override-674gwq ant-btn-default basic-button-wrapper css-geipcv"
                    type="button"
                  >
                    <span>
                      关 闭
                    </span>
                  </button>
                </div>
                <div
                  class="ant-space-item"
                >
                  <button
                    class="ant-btn css-dev-only-do-not-override-674gwq ant-btn-primary basic-button-wrapper css-geipcv"
                    type="button"
                  >
                    <span>
                      提 交
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        aria-hidden="true"
        data-sentinel="end"
        style="width: 0px; height: 0px; overflow: hidden; outline: none; position: absolute;"
        tabindex="0"
      />
    </div>
  </div>
</body>
`;
